File: 2025-and-before/0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

const helper = function(root,obj){
    
    if(!root)
        return;
    
    obj.curr*=10;
    obj.curr+=root.val;

    
    if(!root.left && !root.right){
        obj.ans+=obj.curr;
    }
    
    helper(root.left,obj);
    helper(root.right,obj);
    
    obj.curr-=root.val;
    obj.curr/=10;
    return;
}
var sumNumbers = function(root) {
    
    const obj = {};
    obj.ans = 0;
    obj.curr=0;
    
    helper(root,obj);
    return obj.ans;
    
};

File: 2025-and-before/0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
class Solution{

	private : 
		bool isLeaf(TreeNode *root){
			return (root and !root->left and !root->right);
}
public : 
	
	void helper(TreeNode *root,int&ans,int&currNum){
		if(!root)
		return;
		
		currNum*=10;
		currNum+=root->val;
		
		if(isLeaf(root))
		ans+=currNum;
		
		helper(root->left,ans,currNum);
		helper(root->right,ans,currNum);
		currNum/=10;
		return;
}

	int sumNumbers(TreeNode *root){
		int ans = 0, currNum = 0;
		helper(root,ans,currNum);
        return ans;
}
};
